prober/prober_test.go

```diff
@@ -110,6 +110,32 @@ func TestProberTimingSpread(t *testing.T) {
 	clk.Advance(halfProbeInterval)
 	called()
 	notCalled()
+
+	// We need to wait until the main (non-initial) ticker in Probe.loop is
+	// waiting, or we could race and advance the test clock between when
+	// the initial delay ticker completes and before the ticker for the
+	// main loop is created. In this race, we'd first advance the test
+	// clock, then the ticker would be registered, and the test would fail
+	// because that ticker would never be fired.
+	err := tstest.WaitFor(convergenceTimeout, func() error {
+		clk.Lock()
+		defer clk.Unlock()
+		for _, tick := range clk.tickers {
+			if tick.stopped {
```

Comment:
I think you might also need to lock `tick` here to avoid a data race:

```
==================
WARNING: DATA RACE
Write at 0x00c0002d1fb0 by goroutine 888:
  tailscale.com/prober.(*fakeTicker).Stop()
      /Users/anton/code/tailscale/prober/prober_test.go:332 +0x70
  tailscale.com/prober.(*Probe).loop()
      /Users/anton/code/tailscale/prober/prober.go:156 +0x1a8
  tailscale.com/prober.(*Prober).Run.func2()
      /Users/anton/code/tailscale/prober/prober.go:87 +0x34

Previous read at 0x00c0002d1fb0 by goroutine 887:
  tailscale.com/prober.TestProberTimingSpread.func4()
      /Users/anton/code/tailscale/prober/prober_test.go:124 +0xf8
  tailscale.com/tstest.WaitFor()
      /Users/anton/code/tailscale/tstest/tstest.go:24 +0x108
  tailscale.com/prober.TestProberTimingSpread()
      /Users/anton/code/tailscale/prober/prober_test.go:120 +0x508
  testing.tRunner()
      /Users/anton/.cache/tailscale-go/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /Users/anton/.cache/tailscale-go/src/testing/testing.go:1493 +0x40

Goroutine 888 (running) created at:
  tailscale.com/prober.(*Prober).Run()
      /Users/anton/code/tailscale/prober/prober.go:87 +0x404
  tailscale.com/prober.TestProberTimingSpread()
      /Users/anton/code/tailscale/prober/prober_test.go:100 +0x488
  testing.tRunner()
      /Users/anton/.cache/tailscale-go/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /Users/anton/.cache/tailscale-go/src/testing/testing.go:1493 +0x40
```

Reply:
Good catch; though I didn't repro it in my tests, that's easy to enough to add and seems like a good idea 👍

Reply:
I used `../tool/go test  -race -count 100 -run='^TestProberTimingSpread$'` and the race got detected super quickly, but omitting `-run` makes it harder to repro it seems 🤷